internal/controller/controllers/agentclusterinstall_controller.go

```diff
@@ -0,0 +1,112 @@
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/openshift/assisted-service/internal/common"
+	hiveext "github.com/openshift/assisted-service/internal/controller/api/hiveextension/v1beta1"
+	logutil "github.com/openshift/assisted-service/pkg/log"
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// AgentClusterInstallReconciler reconciles a AgentClusterInstall object
+type AgentClusterInstallReconciler struct {
+	client.Client
+	Log              logrus.FieldLogger
+	CRDEventsHandler CRDEventsHandler
+}
+
+// +kubebuilder:rbac:groups=extensions.hive.openshift.io,resources=agentclusterinstalls,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=extensions.hive.openshift.io,resources=agentclusterinstalls/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=extensions.hive.openshift.io,resources=agentclusterinstalls/finalizers,verbs=update
+
+func (r *AgentClusterInstallReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ctx := addRequestIdIfNeeded(origCtx)
+	log := logutil.FromContext(ctx, r.Log).WithFields(
+		logrus.Fields{
+			"agent_cluster_install":           req.Name,
+			"agent_cluster_install_namespace": req.Namespace,
+		})
+
+	defer func() {
+		log.Info("AgentClusterInstall Reconcile ended")
+	}()
+
+	log.Info("AgentClusterInstall Reconcile started")
+
+	// Retrieve AgentClusterInstall
+	clusterInstall := &hiveext.AgentClusterInstall{}
+	if err := r.Get(ctx, req.NamespacedName, clusterInstall); err != nil {
+		log.WithError(err).Errorf("Failed to get resource %s", req.NamespacedName)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Retrieve ClusterDeployment
+	clusterDeploymentKubeKey := types.NamespacedName{
+		Namespace: clusterInstall.Namespace,
+		Name:      clusterInstall.Spec.ClusterDeploymentRef.Name,
+	}
+	clusterDeployment := &hivev1.ClusterDeployment{}
+	if err := r.Get(ctx, clusterDeploymentKubeKey, clusterDeployment); err != nil {
+		var errMsg string
+		if k8serrors.IsNotFound(err) {
+			apiErr := errors.New(fmt.Sprintf(
+				"ClusterDeployment with name '%s' in namespace '%s' not found. Ensure that the CRD is already applied.",
+				clusterDeploymentKubeKey.Name, clusterDeploymentKubeKey.Namespace))
+			err = common.NewApiError(http.StatusBadRequest, apiErr)
+			log.WithError(err).Error(errMsg)
+		} else {
+			errMsg = fmt.Sprintf("failed to get ClusterDeployment with name '%s' in namespace '%s'",
+				clusterDeploymentKubeKey.Name, clusterDeploymentKubeKey.Namespace)
+			log.WithError(err).Error(errMsg)
+		}
+
+		// Set conditions
+		clusterSpecSynced(clusterInstall, err)
```

Comment:
I'm not sure if it's required to set sync status to error and the conditions to unknown in case of an internal error to get the resource. maybe you just need to requeue? 
@mhrivnak what do you think?

Reply:
So, only update the conditions if we get `IsNotFound`?
Makes sense to me.

Reply:
Sure, added an early return with requeue for other errors.

Reply:
Makes sense to me.